<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Xiaoyao Yin" />

<meta name="date" content="2023-08-11" />

<title>An Introduction to the package M3JF</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">An Introduction to the package M3JF</h1>
<h4 class="author">Xiaoyao Yin</h4>
<h4 class="date">2023-08-11</h4>



<style type="text/css">
body{
font-family: serif;
}
code.r{
font-family: serif;
}
</style>
<p>This vignette presents the <strong>M3JF</strong>,which implements a
framework named multi-modality matrix joint factorization (M3JF) to
conduct integrative analysis of multiple modality data in
<strong>R</strong>. The objective is to provide an implementation of the
proposed method, which is designed to solve the high dimensionality
multiple modality data in bioinformatics. It was achieved by jointly
factorizing the matrices into a shared sub-matrix and several modality
specific sub-matrices. The introduction of group sparse constraint on
the shared sub-matrix forces the samples in the same group to allow each
modality exploiting only a subset of the dimensions of the global latent
space.</p>
<div id="installation" class="section level2">
<h2>Installation</h2>
<p>The latest stable version of the package can be installed from any
CRAN repository mirror:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Install</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&#39;M3JF&#39;</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Load</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(M3JF)</span></code></pre></div>
<p>The latest development version is available from <a href="https://cran.r-project.org/package=M3JF" class="uri">https://cran.r-project.org/package=M3JF</a> and may be
downloaded from there and installed manually:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&#39;/path/to/file/M3JF.tar.gz&#39;</span>,<span class="at">repos=</span><span class="cn">NULL</span>,<span class="at">type=</span><span class="st">&quot;source&quot;</span>)</span></code></pre></div>
<p><strong>Support</strong>: Users interested in this package are
encouraged to email to Xiaoyao Yin (<a href="mailto:xyyin@xmail.ncba.ac.cn" class="email">xyyin@xmail.ncba.ac.cn</a>) for enquiries, bug reports,
feature requests, suggestions or M3JF-related discussions.</p>
</div>
<div id="usage" class="section level2">
<h2>Usage</h2>
<p>We will give an example of how to use this package hereafter.</p>
<div id="simulation-data-generation" class="section level3">
<h3>Simulation data generation</h3>
<p>We generate simulated data with the R package <em>InterSIM</em>,
which generates three inter-related data set with realistic inter- and
intra- relationships based on the DNA methylation, mRNA expression and
protein expression from the TCGA ovarian cancer study. Each data
modality consists of 500 samples, samples are assigned to 4 groups with
100, 150, 135 and 115 samples per group. The data can be generated by
running:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(InterSIM)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>sim.data <span class="ot">&lt;-</span> <span class="fu">InterSIM</span>(<span class="at">n.sample=</span><span class="dv">500</span>, <span class="at">cluster.sample.prop =</span> <span class="fu">c</span>(<span class="fl">0.20</span>,<span class="fl">0.30</span>,<span class="fl">0.27</span>,<span class="fl">0.23</span>),</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>                     <span class="at">delta.methyl=</span><span class="dv">5</span>, <span class="at">delta.expr=</span><span class="dv">5</span>, <span class="at">delta.protein=</span><span class="dv">5</span>,<span class="at">p.DMP=</span><span class="fl">0.2</span>, <span class="at">p.DEG=</span><span class="cn">NULL</span>,</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>                     <span class="at">p.DEP=</span><span class="cn">NULL</span>,<span class="at">sigma.methyl=</span><span class="cn">NULL</span>, <span class="at">sigma.expr=</span><span class="cn">NULL</span>,</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>                     <span class="at">sigma.protein=</span><span class="cn">NULL</span>,<span class="at">cor.methyl.expr=</span><span class="cn">NULL</span>,</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>                     <span class="at">cor.expr.protein=</span><span class="cn">NULL</span>,<span class="at">do.plot=</span><span class="cn">FALSE</span>, <span class="at">sample.cluster=</span><span class="cn">TRUE</span>,</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>                     <span class="at">feature.cluster=</span><span class="cn">TRUE</span>)</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>sim.methyl <span class="ot">&lt;-</span> sim.data<span class="sc">$</span>dat.methyl</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>sim.expr <span class="ot">&lt;-</span> sim.data<span class="sc">$</span>dat.expr</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>sim.protein <span class="ot">&lt;-</span> sim.data<span class="sc">$</span>dat.protein</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>data_list <span class="ot">&lt;-</span> <span class="fu">list</span>(sim.methyl, sim.expr, sim.protein)</span></code></pre></div>
</div>
<div id="simulation-data-groundtruth-assignment-and-permutation" class="section level3">
<h3>Simulation data groundtruth assignment and permutation</h3>
<p><strong>Label assignment</strong>: According to the data generation
process, we assign the groundtruth label to the data we have generated
as:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>truelabel <span class="ot">=</span> sim.data<span class="sc">$</span>clustering.assignment<span class="sc">$</span>cluster.id</span></code></pre></div>
<p>this label will be used to test the clustering ability
afterwards.</p>
<p>Now we can cluster the samples with the proposed method and compare
its performance by calculating the normalized mutual information with
the function <em>cal_NMI</em> by inputting the truelabel and the
predicted label.</p>
<hr />
</div>
</div>
<div id="you-should-start-from-here-if-you-are-using-your-own-data." class="section level2">
<h2>You should start from here if you are using your own data.</h2>
<hr />
<p><strong>Evaluating k</strong>: Evaluate the most proper cluster
number k by mean of modality modulairty with the function
<em>new_modularity</em>.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Build similarity matrices for your data with SNFtool</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SNFtool)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>WL_dist1 <span class="ot">&lt;-</span> <span class="fu">lapply</span>(data_list,<span class="cf">function</span>(x){</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>  dd <span class="ot">&lt;-</span> x<span class="sc">%&gt;%</span>as.matrix</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>  w <span class="ot">&lt;-</span> dd <span class="sc">%&gt;%</span> <span class="fu">dist2</span>(dd) <span class="sc">%&gt;%</span> <span class="fu">affinityMatrix</span>(<span class="at">K =</span> <span class="dv">10</span>, <span class="at">sigma =</span> <span class="fl">0.5</span>)</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="co">#Assign the interval of k according to your data</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>k_list <span class="ot">=</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">10</span></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="co">#Initialize the varible</span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>clu_eval <span class="ot">&lt;-</span> <span class="fu">RotationCostBestGivenGraph</span>(W,k_list)</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a><span class="co">#The most proper is the one with minimal rotation cost</span></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>best_k <span class="ot">=</span> k_list[<span class="fu">which.min</span>(clu_eval)]</span></code></pre></div>
<p><strong>M3JF</strong>: Jointly factorize the matrices into a shared
embedding matrix and several modality private basis matrices.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Assign the parameters</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>lambda <span class="ot">=</span> <span class="fl">0.01</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>theta <span class="ot">=</span> <span class="dv">10</span><span class="sc">^-</span><span class="dv">6</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>k <span class="ot">=</span> best_k</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>res <span class="ot">=</span> <span class="fu">M3JF</span>(data_list,lambda,theta,k)</span></code></pre></div>
<p>Now you have got the classification result you want.</p>
<hr />
</div>
<div id="you-can-ommit-the-following-if-you-do-not-have-any-true-label-as-the-groudtruth-we-do-the-next-to-evaluate-our-method." class="section level2">
<h2>You can ommit the following if you do not have any true label as the
groudtruth, we do the next to evaluate our method.</h2>
<hr />
<p><strong>Robustness test</strong>: We test the robustness of our
method by calculating the normalized mutual information and adjusted
rand index of the true label and our predicted label. We can compare the
performance of our method with others by these scores, which lie in the
interval [0,1]. The larger the scores, the more robust the method.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SNFtool)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="co">#Calculate the NMI of *M3JF*</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>M3JF_res <span class="ot">=</span> <span class="fu">M3JF</span>(data_list,lambda,theta,k)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>M3JF_cluster <span class="ot">=</span> M3JF_res<span class="sc">$</span>clusters</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>M3JF_NMI <span class="ot">=</span> <span class="fu">cal_NMI</span>(true_label,M3JF_cluster)</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="co">#Calculate the ARI of *M3JF*</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mclust)</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>M3JF_ARI <span class="ot">=</span> <span class="fu">adjustedRandIndex</span>(true_label,M3JF_cluster)</span></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
